### PR TITLE
Feature/autoconfig search handshake

### LINKED ISSUE #423: [TASK] Pass flow 4.12.1 test in testbed
# Description

Perform all steps necessary to pass the test associated with flow 4.12.1 in the certification testbed.

# Exit criteria

- Turris Omnia OpenWrt with firmware master branch
- Certification test passes with all configurations
- Test results uploaded to https://ftp.essensium.com/owncloud/index.php/s/ketl3eFhIgwcweZ

### LINKED ISSUE #424: [TASK] Pass flow 4.12.2 test in testbed
# Description

Perform all steps necessary to pass the test associated with flow 4.12.2 in the certification testbed.

# Exit criteria

- Turris Omnia OpenWrt with firmware master branch
- Certification test passes with all configurations
- Test results uploaded to https://ftp.essensium.com/owncloud/index.php/s/ketl3eFhIgwcweZ

### LINKED ISSUE #425: 4.12.1,2 Higher layer data payload over 1905
The prplMesh agent on Turris Omnia with OpenWrt needs to pass the certification test cases 4.12.1 and 4.12.2 as defined by the Wi-Fi CERTIFIED EasyMesh (tm) Test Plan.

- [ ] #423 Pass flow 4.12.1 test in testbed
- [ ] #424 Pass flow 4.12.2 test in testbed

### LINKED ISSUE #675: [TASK] prplMesh autoconfig handshake
## Description

Currently, prplMesh agent and controller always send VS TLVs to some messages even though it is possible the other side will not process these TLVs. Moreover, some testbed certified units don't cope well with fragmented CMDUs (R1 certification has no test for fragmentation/reassembly), so when adding a VS TLV big enough, the CMDU gets fragmented in the transport and then ignored by these units, causing autoconfiguration to fail.

It is possible to avoid this altogether by adding a handshake during the autoconfig search-reply phase - the prplMesh agent will add a VS TLV to the autoconfig search, and the controller will mark this agent as prplMesh if it contains this VS TLV and will answer it with an autoconfig response with a similar VS TLV. The agent will then understand that the controller is also prplMesh, so VS TLVs and VS CMDUs can be used.

In addition, we can use the handshake VS TLVs to add a list of extra supported features that are not part of the specification which can be used in the future by different vendors to differentiate their products.

## Exit criteria

- prplMesh agent always sends autoconfig search with an extra VS TLV with a currently empty list of extra supported features.
- prplMesh controller marks the agent as prplMesh if it sent the autoconfig search with that TLV, and can send VS TLVs and CMDUs to it from now on. Otherwise, it ceases to send VS TLVs to this agent.
-prplMesh agent does the same - if the autoconfig search response contains the VS TLV, it can continue sending VS TLVs and VS CMDUs. Otherwise, it ceases to send those to the controller.

### LINKED ISSUE #921: [TASK] Test onboarding with all controllers
## Description

It seems there was an issue with a test when using a Qualcomm controller, but not with Marvell.
To see if this is a problem with the hardware or with the test, we should really do an onboarding test with all possible configurations, to weed out the various issues.

### PR DESCRIPTION
**Pull request by tomereli**
Currently, prplMesh agent and controller always send VS TLVs to some messages even though it is possible the other side will not process these TLVs. Moreover, some testbed certified units don't cope well with fragmented CMDUs (R1 certification has no test for fragmentation/reassembly), so when adding a VS TLV big enough, the CMDU gets fragmented in the transport and then ignored by these units, causing autoconfiguration to fail.

It is possible to avoid this altogether by adding a handshake during the autoconfig search-reply phase - the prplMesh agent will add a VS TLV to the autoconfig search, and the controller will mark this agent as prplMesh if it contains this VS TLV and will answer it with an autoconfig response with a similar VS TLV. The agent will then understand that the controller is also prplMesh, so VS TLVs and VS CMDUs can be used.

This PR adds the basic handshake using an empty handshake request and reply messages attached to the autoconfig search and response messages.
By doing this, we mark the prplmesh controller as prplmesh or not prplmesh, and can use it to decide if to send VS TLVs and VS messages later on in the agent.

In addition, update the agent's node in the controller DB so that in a future PR we can use it to refrain from sending VS messages to non-prplmesh agents.

Fixes #675 
Fixes #425 
Fixes #423 
Fixes #424 
Fixes #921 

Passing 4.12.1 - https://gitlab.com/prpl-foundation/prplMesh/pipelines/124523357
Passing 4.12.2 - https://gitlab.com/prpl-foundation/prplMesh/-/jobs/464474890

### CODE REVIEW

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml
Comment (vitalii-komisarenko) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml:33]: It breaks backward compartibility.
Are there on the market any older devices?

File: controller/src/beerocks/master/db/db.cpp
Comment (vitalii-komisarenko) [inline controller/src/beerocks/master/db/db.cpp]: Maybe I should have added a debug output here to signal that the node is not in the DB yet and we read `is_prplmesh` too early.

**Review by vitalii-komisarenko — APPROVED** [review]

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml
Comment (tomereli) [reply common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_action.yaml:33]: prplMesh is not deployed it, it is in development, so no.

**Comment by rmelotte** [comment]
Also fixes the issue seen in https://github.com/prplfoundation/prplMesh/issues/921, so I linked it.

**Review by rmelotte — APPROVED** [review]
:rocket: 
I added the don't merge tag just in case you want to squash the commits updating the generated files.

**Comment by tomereli** [comment]
> 🚀
> I added the don't merge tag just in case you want to squash the commits updating the generated files.

Good thing you did - I'd like @arnout to also review this before we merge.

File: controller/src/beerocks/master/db/db.cpp
Comment (tomereli) [reply controller/src/beerocks/master/db/db.cpp]: Fixed.

File: controller/src/beerocks/master/db/db.cpp
Comment (arnout) [inline controller/src/beerocks/master/db/db.cpp]: I *don't* want new interfaces that use strings for MAC addresses.

File: controller/src/beerocks/master/db/db.cpp
Comment (arnout) [inline controller/src/beerocks/master/db/db.cpp]: I don't think a new node should be added in this case. The type of the node is going to be wrong. When we implement #524 we will need to know which type of node to create, so this isn't even possible.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp:2211]: Since we get multiple responses in the backhaul manager, perhaps we should log an error message if we already had prplmesh_controller == true. And we should probably reset it to false as well.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp]: I don't think there's a need to save this flag per socket. We can just use `this->prplmesh_controller` here.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: So I think this member is unnecessary.

File: agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h
Comment (arnout) [inline agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.h]: Nitpick: I prefer names of booleans to be clearly boolean, so `is_prplmesh_controller` or `controller_is_prplmesh`.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml
Comment (arnout) [inline common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:582]: Removal of this unused field should be made explicit in the commit message.

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: Maybe also change this to "non-prplmesh" then?

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (arnout) [inline agent/src/beerocks/slave/son_slave_thread.cpp]: Ideally, we should never even get here... So maybe throw a warning if we do?

File: tests/test_gw_repeater.sh
Comment (arnout) [inline tests/test_gw_repeater.sh:84]: So, this means we no longer have a test of an agent that acts as a non-prplmesh agent. Is that acceptable?

Maybe it's better to keep all this and just force the controller_is_prplmesh option to false if the no_vendor_specific option is set.

**Review by arnout — APPROVED** [review]
Nothing major so approving. Well, I consider the string MAC address a major issue but I assume you can fix it up :-)

There's a typo in one of the commit messages: resonse instead of response.

File: controller/src/beerocks/master/db/db.cpp
Comment (tomereli) [reply controller/src/beerocks/master/db/db.cpp]: I think we create the node only in during the autoconfig M1, so the node will always be created first when we get the autoconfig search request arrives, and if we don't create it here it we won't mark it as prplmesh when we mean to.
@vitalii-komisarenko please confirm.

File: tests/test_gw_repeater.sh
Comment (tomereli) [reply tests/test_gw_repeater.sh:84]: I think it is acceptable since we are testing agent certification on a nightly basis with non-prplmesh controllers and controller certification with non-prplmesh agents.

File: controller/src/beerocks/master/db/db.cpp
Comment (tomereli) [reply controller/src/beerocks/master/db/db.cpp]: Ok.

File: agent/src/beerocks/slave/son_slave_thread.cpp
Comment (tomereli) [reply agent/src/beerocks/slave/son_slave_thread.cpp]: NACK - I did this here instead of not sending the messages and checking if the controller is prplMesh everywhere else where we send the VS message.

File: common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml
Comment (tomereli) [reply common/beerocks/tlvf/yaml/beerocks/tlvf/beerocks_message_common.yaml:582]: Oops I forgot to update the commit message, mergify beat me to it

**Pull request merged**
